### PR TITLE
HHH-14968 - Support for auto-enabled filters

### PR DESCRIPTION
**Pull request by gregoriopalama**
https://hibernate.atlassian.net/browse/HHH-14968

### CODE REVIEW

**Review by gavinking — CHANGES_REQUESTED** [review]
This looks quite good. But I would not introduce the new `FilterParamResolver` interface, since it's isomorphic to `java.util.function.Supplier`.

File: hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java [hunk 1]
```diff
@@ -248,6 +248,14 @@ protected final void setUpMultitenancy(SessionFactoryImplementor factory, LoadQu
 		}
 	}
 
+	protected final void setupAutoEnabledFilters(SessionFactoryImplementor factory, LoadQueryInfluencers loadQueryInfluencers) {
+		factory.getDefinedFilterNames()
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java:252]: Since this is executed whenever a session is created, I would prefer to have `SessionFactoryImplementor` expose a `getAutoEnabledFilters()` that initializes the `enabledFilters` in the constructor of `LoadQueryInfluencers` instead.

File: hibernate-core/src/main/resources/org/hibernate/hibernate-mapping-3.0.dtd [hunk 2]
```diff
@@ -135,6 +135,7 @@ arbitrary number of queries, and import declarations of arbitrary classes.
 <!ELEMENT filter-def (#PCDATA|filter-param)*>
 	<!ATTLIST filter-def name CDATA #REQUIRED> <!-- The filter name -->
 	<!ATTLIST filter-def condition CDATA #IMPLIED>
+	<!ATTLIST filter-def auto-enabled (true|false) "false">
```
Comment (beikov) [inline hibernate-core/src/main/resources/org/hibernate/hibernate-mapping-3.0.dtd:138]: Please revert the changes to this file. hbm.xml is deprecated and will most likely be removed in 7.0.

File: hibernate-core/src/main/resources/org/hibernate/hibernate-mapping-4.0.xsd [hunk 3]
```diff
@@ -65,11 +65,13 @@ arbitrary number of queries, and import declarations of arbitrary classes.
                 <xs:complexType>
                   <xs:attribute name="name" use="required" type="xs:string"/> <!-- The parameter name -->
                   <xs:attribute name="type" use="required" type="xs:string"/> <!-- The parameter type -->
+                  <xs:attribute name="resolver" type="xs:string"/> <!-- The parameter resolver -->
```
Comment (beikov) [inline hibernate-core/src/main/resources/org/hibernate/hibernate-mapping-4.0.xsd:68]: Please revert the changes to this file. hbm.xml is deprecated and will most likely be removed in 7.0.

File: hibernate-core/src/main/resources/org/hibernate/xsd/mapping/legacy-mapping-4.0.xsd [hunk 4]
```diff
@@ -119,11 +119,13 @@
         </xs:sequence>
         <xs:attribute name="condition" type="xs:string"/>
         <xs:attribute name="name" use="required" type="xs:string"/>
+        <xs:attribute name="auto-enabled" type="xs:boolean" default="false"/>
```
Comment (beikov) [inline hibernate-core/src/main/resources/org/hibernate/xsd/mapping/legacy-mapping-4.0.xsd:122]: Please revert the changes to this file. hbm.xml is deprecated and will most likely be removed in 7.0.

File: hibernate-core/src/main/resources/org/hibernate/xsd/mapping/mapping-3.1.0.xsd [hunk 5]
```diff
@@ -2277,12 +2277,14 @@
                     </xsd:annotation>
                     <xsd:attribute name="name" use="required" type="xsd:string"/>
                     <xsd:attribute name="type" use="required" type="xsd:string"/>
+                    <xsd:attribute name="resolver" type="xsd:string"/>
```
Comment (beikov) [inline hibernate-core/src/main/resources/org/hibernate/xsd/mapping/mapping-3.1.0.xsd:2280]: I haven't thought about whether it is sensible to add anything to a versioned XSD. Maybe we should rather create a mapping-3.1.1.xsd? Wdyt @sebersole?

**Comment by gregoriopalama** [comment]
> This looks quite good. But I would not introduce the new `FilterParamResolver` interface, since it's isomorphic to `java.util.function.Supplier`.

Sounds good. I've added the changes

File: hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java [hunk 6]
```diff
@@ -660,11 +661,14 @@ private static void bindFilterDef(FilterDef filterDef, MetadataBuildingContext c
 			throw new AnnotationException( "Multiple '@FilterDef' annotations define a filter named '" + name + "'" );
 		}
 		final Map<String, JdbcMapping> explicitParamJaMappings;
+		final Map<String, Class<? extends Supplier>> parameterResolverMap;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java:664]: I think we should resolve the managed beans already at this point i.e.

```suggestion
		final Map<String, ManagedBean<Supplier>> parameterResolverMap;
```

File: hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/hbm/FilterDefinitionBinder.java [hunk 7]
```diff
@@ -38,6 +39,7 @@ static void processFilterDefinition(
 			HbmLocalMetadataBuildingContext context,
 			JaxbHbmFilterDefinitionType jaxbFilterDefinitionMapping) {
 		Map<String, JdbcMapping> parameterMap = null;
+		Map<String, Class<? extends Supplier>> parameterResolverMap = null;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/hbm/FilterDefinitionBinder.java:42]: I think we should resolve the managed beans already at this point i.e.

```suggestion
		Map<String, ManagedBean<Supplier>> parameterResolverMap = null;
```

File: hibernate-core/src/main/java/org/hibernate/sql/ast/tree/predicate/FilterPredicate.java [hunk 8]
```diff
@@ -36,8 +40,8 @@ public void applyFragment(FilterFragmentPredicate predicate) {
 		fragments.add( predicate );
 	}
 
-	public void applyFragment(String processedFragment, FilterImpl filter, List<String> parameterNames) {
-		fragments.add( new FilterFragmentPredicate( processedFragment, filter, parameterNames ) );
+	public void applyFragment(String processedFragment, FilterImpl filter, List<String> parameterNames, BeanContainer beanContainer) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/tree/predicate/FilterPredicate.java:43]: If we resolve the beans earlier, there is no need for this breaking change.

File: hibernate-core/src/main/java/org/hibernate/sql/ast/tree/predicate/FilterPredicate.java [hunk 9]
```diff
@@ -155,5 +159,36 @@ public JdbcMappingContainer getExpressionType() {
 		public boolean isEmpty() {
 			return false;
 		}
+
+		private Object retrieveParamValue(FilterImpl filter, String paramName, BeanContainer beanContainer) {
+			Class<? extends Supplier> clazz = filter.getParameterResolver( paramName );
+			if (clazz.isInterface()) {
+				return filter.getParameter( paramName );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/tree/predicate/FilterPredicate.java:166]: Regardless of whether there is a parameter resolver, an explicit filter value should IMO always take precedence.

**Review by beikov — CHANGES_REQUESTED** [review]

File: hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java (hunk 6 above)
Comment (gregoriopalama) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java:664]: I was troubled by this, honestly. The first thought was resolving the beans in the AnnotationBinder. I'll do the changes

File: hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java (hunk 1 above)
Comment (gregoriopalama) [reply hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java:252]: How about a `getAutoEnabledFilterNames()`? This way, the constructor of `LoadQueryInfluencers` could enable the filters looping on the names

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java:252]: I would prefer avoiding to do hash map lookups if possible, so I'd rather not expose the names but filter definitions directly.

File: hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java (hunk 6 above)
Comment (gregoriopalama) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java:664]: Just to be sure: if we use ManagedBeans, that will be the only way to use the resolvers: the bean context must be initialized to make the resolvers work. 
The alternative could be sticking to a Map of Suppliers. 
Which way should we go?

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java:664]: >the bean context must be initialized to make the resolvers work.

Not sure what you mean by that exactly. You will just have to call `serviceRegistry.requireService( ManagedBeanRegistry.class ).getBean( supplierClass, context.getBootstrapContext().getCustomTypeProducer() )` to acquire the `ManagedBean`.

Comment (gregoriopalama) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java:664]: Yep, and that's pretty clear. 
As for now, the resolver can be either a ManagedBean or a freshly created object. Tests in FilterParameterTests could clarify this further. I'm ok with sticking to ManagedBeans only, as I'm ok with maintaining the actual options. What I'm asking here is a choice on your side.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java:664]: I'm still unsure what you mean. `ManagedBean` is just an abstraction in Hibernate to abstract over how an instance of a class is acquired. If the bean container does not provide an instance, an instance for the class will be created automatically.

Comment (gregoriopalama) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java:664]: What happens if the `ManagedBeanRegistry` is null, because it has not been initialized?

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java:664]: That should not be possible i.e. `ManagedBeanRegistry` should always be non-null at this point.

Comment (gregoriopalama) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java:664]: Great. Then I'll update the tests too and commit the changes

File: hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java [hunk 10]
```diff
@@ -800,6 +806,15 @@ private static JdbcMapping resolveJavaType(Class<JavaType<?>> type, MetadataBuil
 		return typeConfiguration.getBasicTypeRegistry().resolve( jtd, jdbcType );
 	}
 
+	private static ManagedBean<? extends Supplier> resolveParamResolver(ParamDef paramDef, MetadataBuildingContext context) {
+		Class<? extends Supplier> clazz = paramDef.resolver();
+		if (clazz.isInterface()) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/model/internal/AnnotationBinder.java:811]: It's totally valid to use an interface as bean type. The only case when we should skip trying to create a bean is when the annotation default is used.

```suggestion
		if ( clazz == Supplier.class ) {
```

File: hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/hbm/FilterDefinitionBinder.java [hunk 11]
```diff
@@ -72,19 +75,25 @@ else if ( content instanceof JAXBElement ) {
...
 			}
 		}
 
 		context.getMetadataCollector().addFilterDefinition(
 				new FilterDefinition(
 						jaxbFilterDefinitionMapping.getName(),
 						condition,
-						parameterMap
+						parameterMap,
+						parameterResolverMap,
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/hbm/FilterDefinitionBinder.java:95]: Please revert changes to this file.

File: hibernate-core/src/main/java/org/hibernate/engine/spi/FilterDefinition.java [hunk 12]
```diff
@@ -32,18 +34,25 @@ public class FilterDefinition implements Serializable {
...
+	private final Map<String, ManagedBean<? extends  Supplier>> parameterResolverMap = new HashMap<>();
+	private final boolean autoEnabled;
 
 	/**
 	 * Construct a new FilterDefinition instance.
 	 *
 	 * @param name The name of the filter for which this configuration is in effect.
 	 */
-	public FilterDefinition(String name, String defaultCondition, @Nullable Map<String, JdbcMapping> explicitParamJaMappings) {
+	public FilterDefinition(String name, String defaultCondition, @Nullable Map<String, JdbcMapping> explicitParamJaMappings,
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/engine/spi/FilterDefinition.java:45]: Please re-add a construct `public FilterDefinition(String name, String defaultCondition, @Nullable Map<String, JdbcMapping> explicitParamJaMappings) {` that delegates to this constructor, additionally passing `Collections.emptyMap(), false`

File: hibernate-core/src/main/java/org/hibernate/engine/spi/FilterDefinition.java [hunk 13]
```diff
@@ -32,18 +34,25 @@ public class FilterDefinition implements Serializable {
...
+							@Nullable Map<String, ManagedBean<? extends  Supplier>> parameterResolverMap, boolean autoEnabled) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/engine/spi/FilterDefinition.java:46]:
```suggestion
							Map<String, ManagedBean<? extends Supplier>> parameterResolverMap, boolean autoEnabled) {
```

File: hibernate-core/src/main/java/org/hibernate/engine/spi/FilterDefinition.java [hunk 14]
```diff
@@ -32,18 +34,25 @@ public class FilterDefinition implements Serializable {
 	private final String filterName;
 	private final String defaultFilterCondition;
 	private final Map<String, JdbcMapping> explicitParamJaMappings = new HashMap<>();
+	private final Map<String, ManagedBean<? extends  Supplier>> parameterResolverMap = new HashMap<>();
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/engine/spi/FilterDefinition.java:37]:
```suggestion
	private final Map<String, ManagedBean<? extends Supplier>> parameterResolverMap = new HashMap<>();
```

File: hibernate-core/src/main/java/org/hibernate/engine/spi/FilterDefinition.java [hunk 15]
```diff
@@ -77,6 +86,13 @@ public Set<String> getParameterNames() {
 		return explicitParamJaMappings.get( parameterName );
 	}
 
+	public @Nullable Supplier getParameterResolver(String parameterName) {
+		if ( !parameterResolverMap.containsKey(parameterName) ) {
+			return null;
+		}
+		return parameterResolverMap.get( parameterName ).getBeanInstance();
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/engine/spi/FilterDefinition.java:93]:
```suggestion
		final ManagedBean<? extends Supplier> resolver = parameterResolverMap.get( parameterName );
		return resolver == null ? null : resolver.getBeanInstance();
```

File: hibernate-core/src/test/java/org/hibernate/orm/test/filter/FilterParameterTests.java [hunk 16]
```diff
@@ -182,22 +195,131 @@ public void testMismatch(BiConsumer<SessionFactoryScope, Consumer<? extends Shar
...
 	public void dropTestData() {
 		scope.inTransaction( (session) -> {
+			session.disableFilter( "subDepartmentFilter" );
+			session.disableFilter( "departmentFilter" );
 			session.createMutationQuery( "delete EntityOne" ).executeUpdate();
 			session.createMutationQuery( "delete EntityTwo" ).executeUpdate();
 			session.createMutationQuery( "delete EntityThree" ).executeUpdate();
+			session.createMutationQuery( "delete EntityFour" ).executeUpdate();
+			session.enableFilter( "subDepartmentFilter" );
+			session.enableFilter( "departmentFilter" );
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/filter/FilterParameterTests.java:322]: Shouldn't be necessary to re-enable the filters.

```suggestion
```

**Review by beikov — CHANGES_REQUESTED** [review]

File: hibernate-core/src/main/java/org/hibernate/sql/ast/tree/predicate/FilterPredicate.java [hunk 17]
```diff
@@ -155,5 +159,15 @@ public JdbcMappingContainer getExpressionType() {
...
 		}
+
+		private Object retrieveParamValue(FilterImpl filter, String paramName) {
+			Object value = filter.getParameter(paramName);
+			if (value != null) {
+				return value;
+			}
+
+			final Supplier filterParamResolver = filter.getParameterResolver( paramName );
+			return filterParamResolver.get() == null ? null : filterParamResolver.get();
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/tree/predicate/FilterPredicate.java:170]:
```suggestion
			return filterParamResolver == null ? null : filterParamResolver.get();
```

File: hibernate-core/src/main/java/org/hibernate/internal/SessionFactoryImpl.java [hunk 18]
```diff
@@ -260,6 +261,11 @@ public SessionFactoryImpl(
 			//noinspection unchecked
 			tenantIdentifierJavaType = jdbcMapping.getJavaTypeDescriptor();
 		}
+		for (Map.Entry<String, FilterDefinition> filterEntry : filters.entrySet()) {
+			if (filterEntry.getValue().isAutoEnabled()) {
+				autoEnabledFilters.add( filterEntry.getValue() );
+			}
+		}
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/internal/SessionFactoryImpl.java:268]:
```suggestion
		autoEnabledFilters = new HashSet<>( filters.values() );
```

**Review by beikov — CHANGES_REQUESTED** [review]
Almost done ;)

File: hibernate-core/src/main/java/org/hibernate/internal/SessionFactoryImpl.java (hunk 18 above)
Comment (gregoriopalama) [reply hibernate-core/src/main/java/org/hibernate/internal/SessionFactoryImpl.java:268]: Shouldn't `autoEnabledFilters` contains only the `FilterDefinition`s that have the autoEnabled flag set to true? With this suggestion, we wouldn't filter them anymore

File: hibernate-core/src/main/resources/org/hibernate/xsd/mapping/mapping-3.1.0.xsd (hunk 5 above)
Comment (sebersole) [reply hibernate-core/src/main/resources/org/hibernate/xsd/mapping/mapping-3.1.0.xsd:2280]: As we discussed on the call, let's change to base this version on Hibernate version rather than JPA version

Comment (beikov) [reply hibernate-core/src/main/resources/org/hibernate/xsd/mapping/mapping-3.1.0.xsd:2280]: After our team meeting, we decided to create a new mapping file per ORM major/minor version, so this change would have to go to the new file. @sebersole is going to commit these changes later today I hope, then you can rebase. After that, I think this is good to go.

Comment (gregoriopalama) [reply hibernate-core/src/main/resources/org/hibernate/xsd/mapping/mapping-3.1.0.xsd:2280]: @beikov @sebersole any update on this?

**Comment by kingan379** [comment]
Any updates when it will be merged?

File: hibernate-core/src/main/resources/org/hibernate/xsd/mapping/mapping-3.1.0.xsd (hunk 5 above)
Comment (sebersole) [reply hibernate-core/src/main/resources/org/hibernate/xsd/mapping/mapping-3.1.0.xsd:2280]: I've already started this on the 7.0 branch.  

I am getting ready to release 6.5, so I took your branch and and making the change on top of that.  I'll send a PR in a few

**Comment by sebersole** [comment]
I pulled this locally and fixed a conflict.  Pushed upstream as https://github.com/hibernate/hibernate-orm/commit/527beb0bdb0ca65bba076e800ea55271abe972f8

File: hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java [hunk 19]
```diff
@@ -248,6 +248,14 @@ protected final void setUpMultitenancy(SessionFactoryImplementor factory, LoadQu
 		}
 	}
 
+	protected final void setupAutoEnabledFilters(SessionFactoryImplementor factory, LoadQueryInfluencers loadQueryInfluencers) {
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java:251]: I assume that this method should have been deleted from the pull request, since it currently has no callers.

@beikov @gregoriopalama
